backend/auth/firebase.go

```diff
@@ -16,55 +16,92 @@ package auth
 
 import (
 	"context"
-	"errors"
 	"firebase.google.com/go/v4/auth"
 	"net/http"
+	"peoplemath/models"
 	"strings"
 	"time"
 )
 
 type FirebaseAuth struct {
 	FirebaseClient firebaseAuthClient
 	AuthTimeout    time.Duration
-	AuthDomain     string
 }
 
 type firebaseAuthClient interface {
 	VerifyIDToken(ctx context.Context, idToken string) (*auth.Token, error)
 }
 
-func (auth *FirebaseAuth) Authorize(next http.HandlerFunc) http.HandlerFunc {
+func (auth FirebaseAuth) Authenticate(next http.HandlerFunc) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
+		ctx, cancel := context.WithTimeout(r.Context(), auth.AuthTimeout)
+		defer cancel()
+
 		idToken := strings.TrimPrefix(r.Header.Get("Authorization"), "Bearer ")
-		userEmail, err := auth.Authenticate(r.Context(), idToken)
+		token, err := auth.FirebaseClient.VerifyIDToken(ctx, idToken)
 		if err != nil {
 			w.Header().Add("Authorization", "WWW-Authenticate: Bearer")
-			http.Error(w, err.Error(), http.StatusUnauthorized)
+			http.Error(w, "User authentication failed", http.StatusUnauthorized)
 			return
 		}
-		if getDomain(userEmail) != auth.AuthDomain {
-			http.Error(w, "You are not authorized to view this resource", http.StatusForbidden)
+		if !token.Claims["email_verified"].(bool) {
+			w.Header().Add("Authorization", "WWW-Authenticate: Bearer")
+			http.Error(w, "User authentication failed, please verify your email address", http.StatusUnauthorized)
 			return
 		}
-		next(w, r)
+
+		userEmail := token.Claims["email"].(string)
+		user := User{
+			email:  userEmail,
+			domain: getDomain(userEmail),
+		}
+
+		ctxWithUser := context.WithValue(r.Context(), "user", user)
+		next(w, r.WithContext(ctxWithUser))
 	}
 }
 
-func (auth *FirebaseAuth) Authenticate(ctx context.Context, idToken string) (userEmail string, err error) {
-	ctx, cancel := context.WithTimeout(ctx, auth.AuthTimeout)
-	defer cancel()
+func getPermissionsFromTeam(team models.Team, action string) []models.Principal {
```

Comment:
We should come up with a better name for these. Maybe `getTeamAllowed` etc?

Reply:
I find getTeamAllowed a bit confusing because it's not clear what is being returned, I've changed it to getTeamAllowedUsers now which would obviously really be get team allowed user matchers but I did want to keep the function name shorter, let me know what you think.

Reply:
Thanks, this is fine.